File: lexer/src/main/java/net/jr/lexer/Terminal.java
package net.jr.lexer;

import net.jr.common.Symbol;

/**
 * A {@link Terminal} is the basicterminals unit of meaning in a grammar. It is a synonymn of "terminal symbol"
 */
public interface Terminal extends Symbol {

    @Override
    default boolean isTerminal() {
        return true;
    }

    int getPriority();

    void setPriority(int priority);

    Terminal withPriority(int priority);

    String getName();

    void setName(String name);

    default Terminal withName(String name) {
        setName(name);
        return this;
    }
}

File: src/common/Terminal.ts
import { ParseSymbol } from './ParseSymbol';
import { Automaton } from '../lexer/automaton/Automaton';

export abstract class Terminal extends ParseSymbol {

    public priority:number = 0;

    constructor(name: string) {
        super(name);
    }

    isTerminal() {
        return true;
    }

    abstract get automaton(): Automaton;
};